### PR TITLE
Add Python 3 support to deployd

### PR DESCRIPTION
**Pull request by jogo**
Add python 3 support in a backwards compatible manor so Python 2 still works by using [futurize](https://python-future.org/).

Make sure all unit tests pass via running 'tox'

### CODE REVIEW

File: deploy-agent/deployd/client/client.py [hunk 1]
```diff
@@ -175,8 +175,8 @@ def send_reports(self, env_reports=None):
             if self._read_host_info():
                 reports = [status.report for status in env_reports.values()]
                 for report in reports:
-                    if isinstance(report.errorMessage, str):
-                        report.errorMessage = unicode(report.errorMessage, "utf-8")
+                    if isinstance(report.errorMessage, bytes):
```
Comment (yongwen) [inline deploy-agent/deployd/client/client.py:178]: will this work for both py2 and py3?  surprising to see the type "str" and "bytes" are the same

File: deploy-agent/deployd/common/config.py [hunk 2]
```diff
@@ -14,14 +15,16 @@
...
 import os
-from ConfigParser import SafeConfigParser
 
 from deployd.common.exceptions import DeployConfigException
 from deployd.common.types import DeployType
 from deployd.common.utils import exit_abruptly
 from deployd.types.deploy_stage import DeployStage
 from deployd.types.opcode import OperationCode
 
+from configparser import ConfigParser
```
Comment (yongwen) [inline deploy-agent/deployd/common/config.py:25]: just double check it is also supported in py2 and it is a safe.

File: deploy-agent/deployd/common/env_status.py [hunk 3]
```diff
@@ -42,13 +42,13 @@ def load_envs(self):
...
-                envs = {key: DeployStatus(json_value=d) for key, d in data.iteritems()}
+                envs = {key: DeployStatus(json_value=d) for key, d in data.items()}
         except IOError:
             log.info("Could not find file {}. It happens when run deploy-agent the "
                      "first time, or there is no deploy yet.".format(self._status_fn))
             return {}
-        except Exception as e:
-            log.error(e.message)
+        except Exception:
+            log.exception("Something went wrong in load_envs")
```
Comment (yongwen) [inline deploy-agent/deployd/common/env_status.py:51]: would be nice to log the exception messsage as the original intent

File: deploy-agent/deployd/common/utils.py [hunk 4]
```diff
@@ -45,7 +46,7 @@ def exit_abruptly(status=0):
 
 def touch(fname, times=None):
     try:
-        with file(fname, 'a'):
+        with open(fname, 'a'):
```
Comment (yongwen) [inline deploy-agent/deployd/common/utils.py:49]: can we still use "file" in py3? assuming "open" will actually open the file handle and "file" does not.

File: deploy-agent/requirements.txt [hunk 5]
```diff
@@ -3,10 +3,9 @@
 # tests
 tox==1.6.1
 coverage
-pytest==2.3.5
+pytest
```
Comment (yongwen) [inline deploy-agent/requirements.txt:6]: it is better to pin down a version

File: deploy-agent/setup.py [hunk 6]
```diff
@@ -26,10 +26,12 @@
 
 install_requires = [
     "requests==2.20.0",
-    "gevent>=1.0.2,<=1.2.2",
+    "gevent>=1.0.2,<=1.2.2; python_version < '3'",
+    "gevent>=1.0.2,<=1.5.0; python_version >= '3'",
```
Comment (yongwen) [inline deploy-agent/setup.py:30]: in the past,  upgrading gevent caused issues. just need to do more testing on the new version of gevent in py3.

**Review by yongwen — COMMENTED** [review]
Thanks Joe.

File: deploy-agent/deployd/client/client.py (hunk 1 above)
Comment (jogo) [reply deploy-agent/deployd/client/client.py:178]: yes, under PY2 `bytes==str`

File: deploy-agent/deployd/common/config.py (hunk 2 above)
Comment (jogo) [reply deploy-agent/deployd/common/config.py:25]: yup confirmed by running unit tests work on PY2 and PY3

File: deploy-agent/deployd/common/env_status.py (hunk 3 above)
Comment (jogo) [reply deploy-agent/deployd/common/env_status.py:51]: log.exception already does that

File: deploy-agent/deployd/common/utils.py (hunk 4 above)
Comment (jogo) [reply deploy-agent/deployd/common/utils.py:49]: file==open 

https://portingguide.readthedocs.io/en/latest/builtins.html#removed-file

https://docs.python.org/2.7/library/functions.html#file

File: deploy-agent/requirements.txt (hunk 5 above)
Comment (jogo) [reply deploy-agent/requirements.txt:6]: for testing only I would rather use the latest version, but happy to pin to a newer version of pytest if you want

File: deploy-agent/setup.py (hunk 6 above)
Comment (jogo) [reply deploy-agent/setup.py:30]: note the update is only for PY3 and not PY2.  Do you have any links to previous issues?

File: deploy-agent/requirements.txt (hunk 5 above)
Comment (yongwen) [reply deploy-agent/requirements.txt:6]: can you pin down to a latest version? i think it is better to have deterministic test

Comment (jogo) [reply deploy-agent/requirements.txt:6]: I don't think it is actually, we don't pin coverage either

Comment (yongwen) [reply deploy-agent/requirements.txt:6]: i think we should always pin down to a version for pip. we are not using coverage, so it is ok to remove this dependency all together.

**Review by yongwen — APPROVED** [review]

**Pull request merged**